### PR TITLE
Hanging protocol visualization

### PR DESCRIPTION
**Pull request by amickan**
This adds a visualization of the hanging protocol to the create and update form as well as to the detail view and the list view. 

image

image

https://user-images.githubusercontent.com/30069334/161499758-123ce6f8-015a-40a3-993c-ea22cb7bc464.mp4

### CODE REVIEW

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    let allW = [];
+    let allH = [];
+
+    for (let i = 0; i < json.length; i++) {
+        json[i].x == undefined ? undefined : allX.push(json[i].x);
+        json[i].y == undefined ? undefined : allY.push(json[i].y);
+        json[i].w == undefined ? undefined : allW.push(json[i].w);
+        json[i].h == undefined ? undefined : allH.push(json[i].h);
+    }
+    let totalWidth = (allX.length != json.length | allW.length != json.length) ? json.length : Math.max(...allX.map(function(num, idx) {return num+allW[idx]}));
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:31]: The `|` in the first part of the ternary operator here is a bitwise OR. It think you meant to use the logical OR (=`||` in JS `or` in Python). Functionally it will not make a difference in this case since you apply it on booleans (except for very minor performance increase due to short circuiting) but you should always use the logical OR in cases like these.


Also, the logic on these lines will fail silently if there are multiple viewports defined that only contain some of the `x`, `y`, `w` or `h` fields. I think it is probably best to add an extra requirement to the hanging protocol that either all of these fields for all viewports are filled in, or none of them. That way it will be a lot easier to calculate max heigh/width and it will be a lot clearer also for the users. Then this whole function can also be simplified.

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    return [totalHeight, totalWidth]
+}
+
+function createViewportDiv(divId, viewportNum, viewportSpec, totalHeight, totalWidth) {
+    let viewportDiv = document.createElement("div");
+    viewportDiv.setAttribute('id', viewportSpec.viewport_name);
+    viewportDiv.style.opacity = '0.5';
+    viewportDiv.style.position = 'absolute';
+    viewportDiv.style.zIndex = '-' + viewportSpec.order;
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:42]: Nice!

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    viewportDiv.style.width = isNaN(viewportSpec.w) ? (1 / parseFloat(totalWidth).toFixed(2))*100 + '%' : (viewportSpec.w / parseFloat(totalWidth).toFixed(2))*100 + '%';
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:43]: The `toFixed()` turns a Number into a string with a certain number of decimals. And javascript tries to change the type of a certain variable so that the operation applied to it makes sense (type coercion). So what will happen when the interpreter comes across this statement (notice the quotes to indicate that it is a string):
```
> const totalWidth = "123.456"
undefined
> parseFloat(totalWidth)
123.456
> parseFloat(totalWidth).toFixed(2)
'123.46'
> 1 / parseFloat(totalWidth).toFixed(2)
0.008099789405475458
> (1 / parseFloat(totalWidth).toFixed(2)) * 100
0.8099789405475458
> (1 / parseFloat(totalWidth).toFixed(2)) * 100 + '%'
'0.8099789405475458%'
```

So that is probably not quite what you wanted to do. I think you wanted to do something like this:
```
(1 / parseFloat(totalWidth) * 100).toFixed(2) + '%'
```
And even here the `parseFloat` is not strictly necessary because the interpreter will already try to convert the string `totalWidth` into a Number when you try to apply the divide operation on it, but it is clearer to add it.

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            createViewportDiv(parentDivId, i, jsonSpec[i], totalHeight, totalWidth);
+        }
+    } catch (err) {
+    }
+}
+
+$(document).ready(function () {
+    if (window.location.href.includes("update") | window.location.href.includes("create")) {
+        let jsonString = document.getElementById("id_json").value;
+        showOrHideVisualizationDiv("hpVisualization", jsonString);
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:77]: This line is not necessary since it is already called in the `updateHangingProtocolVisualization` function below, right?

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    let totalWidth = (allX.length != json.length | allW.length != json.length) ? json.length : Math.max(...allX.map(function(num, idx) {return num+allW[idx]}));
+    let totalHeight = (allY.length != json.length | allH.length != json.length) ? 1 : Math.max(...allY.map(function(num, idx) {return num+allH[idx]}));
+
+    return [totalHeight, totalWidth]
+}
+
+function createViewportDiv(divId, viewportNum, viewportSpec, totalHeight, totalWidth) {
+    let viewportDiv = document.createElement("div");
+    viewportDiv.setAttribute('id', viewportSpec.viewport_name);
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:39]: If `viewportSpec.viewport_name` is not (yet) defined, this will create a div with `id="undefined"`. So if there are multiple empty objects there will be multiple divs with `id="undefined"` but id should always be unique. Also it will show undefined as the name of the div. I think it would be nicer if either the div is not created when the `viewport_name`  field is not (yet) defined, or that it will show no title and for example a red background or something to indicate that it is an invalid viewport.

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    viewportDiv.style.opacity = '0.5';
+    viewportDiv.style.position = 'absolute';
+    viewportDiv.style.zIndex = '-' + viewportSpec.order;
+    viewportDiv.style.width = isNaN(viewportSpec.w) ? (1 / parseFloat(totalWidth).toFixed(2))*100 + '%' : (viewportSpec.w / parseFloat(totalWidth).toFixed(2))*100 + '%';
+    viewportDiv.style.height = isNaN(viewportSpec.h) ? "100%" : (viewportSpec.h / parseFloat(totalHeight).toFixed(2))*100 + '%';
+    viewportDiv.style.left = isNaN(viewportSpec.x) ? (viewportNum / parseFloat(totalWidth).toFixed(2))*100 + '%' : (viewportSpec.x / parseFloat(totalWidth).toFixed(2))*100 + '%';
+    viewportDiv.style.top = isNaN(viewportSpec.y) ? "0%" : (viewportSpec.y / parseFloat(totalHeight).toFixed(2))*100 + '%';
+    document.getElementById(divId).appendChild(viewportDiv).classList.add('bg-dark', 'rounded', 'border', 'border-2', 'd-flex', 'flex-column', 'justify-content-center', 'align-items-center');
+
+    if (/\d/.test(divId)) {
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:49]: If I understand correctly, this test is to see if the parent div has a digit in it so if you are on the list view, right? I think this should be done in a more robust way. It would be even better if this is not needed at all. I think you can do that by making the font size relative and always show the icon for fullsizable.

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        removeAllChildNodes(document.getElementById(parentDivId));
+        for (let i = 0; i < jsonSpec.length; i++) {
+            createViewportDiv(parentDivId, i, jsonSpec[i], totalHeight, totalWidth);
+        }
+    } catch (err) {
+    }
+}
+
+$(document).ready(function () {
+    if (window.location.href.includes("update") | window.location.href.includes("create")) {
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:75]: This works but there should be a better way to check if the user is on a page where they are editing the hp form. I think the best way is to remove it from this file and add it to a separate js file that you only include in the form.html page.

**Review by HarmvZ — CHANGES_REQUESTED** [review]
Nice work on this! I like the minimal styling of it and that you also added it to the detail and list views. I have a few comments about the js stuff but other than that it looks great!

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js (hunk 5 above)
Comment (amickan) [reply app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:39]: Good point. I went with a mix of the two. The div only gets created when the `viewport_name` is not undefined, and if the provided `viewport_name` is not one of the possible labels, it gets a red background and a warning that a valid viewport name needs to be selected.

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    viewportDiv.style.height = isNaN(viewportSpec.h) ? "100%" : (viewportSpec.h / parseFloat(totalHeight) * 100).toFixed(2) + '%';
+    viewportDiv.style.left = isNaN(viewportSpec.x) ? (viewportNum / parseFloat(totalWidth) * 100).toFixed(2) + '%' : (viewportSpec.x / parseFloat(totalWidth) * 100).toFixed(2) + '%';
+    viewportDiv.style.top = isNaN(viewportSpec.y) ? "0%" : (viewportSpec.y / parseFloat(totalHeight) * 100).toFixed(2) + '%';
+    document.getElementById(divId).appendChild(viewportDiv).classList.add('rounded', 'border', 'border-2', 'd-flex', 'flex-column', 'justify-content-center', 'align-items-center');
+
+    if (!possibleViewPorts.includes(viewportSpec.viewport_name)) {
+        viewportDiv.innerHTML += '<p class="mb-0" style="color: #fff">Invalid viewport name</p>';
+    } else {
+        viewportDiv.innerHTML += '<p class="mb-0">' + viewportSpec.viewport_name + '</p>';
+        if (viewportSpec.fullsizable == true) {
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:60]: I missed this on my first passthrough, but it is best to never use double equals in JS because that can lead to weird behavior. So instead of using `==` and `!=` you should always use `===` and `!==`. In this case it's just checking if the value is true so you can simplify to:

```suggestion
        if (viewportSpec.fullsizable) {

```

File: app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    }
+}
+
+function updateHangingProtocolVisualization(parentDivId, jsonString){
+    jsonString = jsonString || document.getElementById("id_json").value;
+    parentDivId = parentDivId || "hpVisualization";
+    showOrHideVisualizationDiv(parentDivId, jsonString);
+    try {
+        let jsonSpec = JSON.parse(jsonString);
+        let validJsonSpec = jsonSpec.filter(viewPort => viewPort.viewport_name != undefined);
```
Comment (HarmvZ) [inline app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js:72]: Never use `==`. In this case because you are checking if the value is undefined it gets slightly more complex. The way to do that properly is like this:

```suggestion
        let validJsonSpec = jsonSpec.filter(viewPort => typeof viewPort.viewport_name !== "undefined");

```

File: app/tests/hanging_protocols_tests/test_forms.py [hunk 10]
```diff
@@ -102,3 +102,47 @@ def test_hanging_protocol_form(client):
...
+        follow=True,
+        user=user,
+    )
+
+    assert response.status_code == 200
+    assert HangingProtocol.objects.count() == 1
+    assert (
+        "Viewport definition must contain all or none of x, y, w, and h"
+        not in response.rendered_content
+    )
```
Comment (HarmvZ) [inline app/tests/hanging_protocols_tests/test_forms.py:148]: Maybe add a test for the case where there are multiple viewports and one contains all of the dims and the other none resulting in an error.

**Review by HarmvZ — CHANGES_REQUESTED** [review]
I have a few comments left. I missed some small things on the first review (`==` vs `===`) and I messed up the example I gave you for the form validation. Sorry 😅

**Review by HarmvZ — APPROVED** [review]
